### PR TITLE
Add Deltalake-Spark3 Demo

### PR DESCRIPTION
**Pull request by snazy**
The nessiedemo-library code for the Delta-Spark demo uses a very similar approach as
the existing Iceberg-Spark nessiedemo-code. So I've refactored the related nessiedemo-library
code to a base class that concrete classes for Iceberg-Spark and Delta-Spark use.

The existing notebook `nessie-iceberg-demo-nba.ipynb` has not been renamed, as that
would break existing links to that demo.

This PR also removes the `TypeDef/T` code.

Configuration file and test-code for Delta-Spark2 is there, but disabled, because there
are issues with Delta < 0.8.

This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie-demos/31)

### CODE REVIEW

**Comment by snazy** [comment]
Note: the new notebook and it's test doesn't work yet, it's derfinitely still lacking a bunch of stuff

**Comment by snazy** [comment]
This one is nearly ready. It currently requires a custom Nessie build with https://github.com/projectnessie/nessie/pull/1249, so as soon as that Nessie PR is merged and a corresponding Nessie version released, this PR can be finalized and merge.

File: colab/nessie-delta-demo-nba.ipynb [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+    "jupyter": {
+     "outputs_hidden": false
+    },
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# `main` hasn't changed been changed and still has the `totals_stats` table\n",
```
Comment (nastra) [inline colab/nessie-delta-demo-nba.ipynb:441]: `main` hasn't been changed and still has the `totals_stats` table

File: colab/nessie-delta-demo-nba.ipynb [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+    "!nessie contents --list"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's take a look at the contents of the `salaries` table on the `experiment` branch.\n",
+    "Notice the use of the `nessie` catalog and the use of `@experiment` to view data on the `experiment` branch"
```
Comment (nastra) [inline colab/nessie-delta-demo-nba.ipynb:450]: this probably should be removed then

File: colab/nessie-delta-demo-nba.ipynb [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+    "# Switch to the `experiment` branch.\n",
+    "demo_delta_spark.change_ref(\"experiment\")\n",
+    "spark.sql(\"select count(*) from nessie_nba_salaries\").show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "and compare to the contents of the `salaries` table on the `main` branch. Notice that we didn't have to specify `@branchName` as it defaulted\n",
```
Comment (nastra) [inline colab/nessie-delta-demo-nba.ipynb:468]: same here, we're mentioning `@branchName` usage

**Review by nastra — CHANGES_REQUESTED** [review]

File: colab/nessie-delta-demo-nba.ipynb [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Notice how we view the data of the salaries table on the `dev` branch via `@dev` via the\n",
+    "# `spark` session, which itself points to the `main` branch.\n",
+    "# We could get the same result using the `spark_dev` session by executing\n",
+    "# spark_dev.sql(\"select * from nessie.nba.salaries\").show()\n",
+    "\n",
+    "spark.sql(\"select * from nessie_nba_salaries\").show()"
```
[the remark this answers is not in the document]
Comment (snazy) [reply colab/nessie-delta-demo-nba.ipynb:145]: Delta doesn't support the `@reference` syntax at all, sad.

**Review by nastra — APPROVED** [review]

**Pull request merged**